src/analysis/normed_space/box_subadditive.lean

```diff
@@ -0,0 +1,541 @@
+box_subsupadditive_coe_helper (≥) (≥) (λ _ _, nnreal.coe_le_coe) nnreal.coe_add
+
+end coe
+
+/-!
+### Examples of `box_additive`, `box_subadditive, and `box_supadditive` functions
+-/
+
+section
+
+lemma box_additive_on_prod_sub [decidable_eq ι] [fintype ι] (s : set (ι → ℝ)) :
+  box_additive_on (λ l r, ∏ i, (r i - l i)) s :=
+box_additive_on.prod (λ x y : ℝ, y - x) $ λ x y z _ _, sub_add_sub_cancel' _ _ _
+
+lemma box_additive_on_prod_dist [decidable_eq ι] [fintype ι] (s : set (ι → ℝ)) :
+  box_additive_on (λ l r, ∏ i, dist (l i) (r i)) s :=
+by simpa only [real.dist_eq, abs_prod, abs_sub]
+  using (box_additive_on_prod_sub s).abs_of_nonneg
+    (λ l u h _, prod_nonneg (λ i _, sub_nonneg.2 (h _)))
+
+lemma box_additive_on_prod_nndist [decidable_eq ι] [fintype ι] (s : set (ι → ℝ)) :
+  box_additive_on (λ l r, ∏ i, nndist (l i) (r i)) s :=
+by simpa only [← box_additive_on.coe_nnreal, nnreal.coe_prod, coe_nndist]
+  using box_additive_on_prod_dist s
+
+lemma box_additive_on_prod_edist [decidable_eq ι] [fintype ι] (s : set (ι → ℝ)) :
+  box_additive_on (λ l r, ∏ i, edist (l i) (r i)) s :=
+by simpa only [edist_nndist, ← ennreal.coe_finset_prod, box_additive_on.coe_ennreal]
+  using box_additive_on_prod_nndist s
+
+end
+
+namespace box_subadditive_on
+
+section preorder
+
+variables [decidable_eq ι] [fintype ι] [preorder α]
+  {s : set (ι → α)} {f g : (ι → α) → (ι → α) → ennreal} {l m u : ι → α}
+
+lemma exists_subbox_mul_lt_of_mul_lt (hf : box_subadditive_on f s) (hg : box_supadditive_on g s)
+  (hsub : Icc l u ⊆ s) (hm : m ∈ Icc l u) {c : ennreal} (hlt : c * g l u < f l u) :
+  ∃ t : finset ι,
+    c * g (t.piecewise m l) (t.piecewise u m) < f (t.piecewise m l) (t.piecewise u m) :=
+begin
+  contrapose! hlt,
+  calc f l u ≤ ∑ t : finset ι, f (t.piecewise m l) (t.piecewise u m) : hf.le_sum_subboxes hsub hm
+  ... ≤ ∑ t : finset ι, c * g (t.piecewise m l) (t.piecewise u m) : sum_le_sum (λ t _, hlt t)
+  ... = c * ∑ t : finset ι, g (t.piecewise m l) (t.piecewise u m) : mul_sum.symm
+  ... ≤ c * g l u : canonically_ordered_semiring.mul_le_mul_left' (hg.sum_subboxes_le hsub hm) c
+end
+
+end preorder
+
+variables [decidable_eq ι] [fintype ι]
+
+noncomputable theory
+
+variables {s : set (ι → ℝ)} {l u : ι → ℝ}
+
+section ennreal
+
+variables {f g : (ι → ℝ) → (ι → ℝ) → ennreal} {c : ennreal}
+
+/-- An auxiliary type for the proof of `box_subadditive_on.eq_zero_of_forall_eventually_le_mul`. -/
+@[nolint has_inhabited_instance]
+structure subbox_mul_lt (s : set (ι → ℝ)) (f g : (ι → ℝ) → (ι → ℝ) → ennreal) (c : ennreal) :=
+(left right : ι → ℝ)
+(le : left ≤ right)
+(sub : Icc left right ⊆ s)
+(mul_lt : c * g left right < f left right)
+
+lemma subbox_mul_lt.midpoint_mem (I : subbox_mul_lt s f g c) :
+  midpoint ℝ I.left I.right ∈ Icc I.left I.right :=
+⟨left_le_midpoint.2 I.le, midpoint_le_right.2 I.le⟩
+
+/-- An auxiliary definition for `box_subadditive_on.eq_zero_of_forall_eventually_le_mul`. -/
```

Comment:
This docstring does not really help. Can you explain what it does, and why you need it?